doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,483 @@
+
+Moreover, our QUIC implementation will not drive the Berkeley sockets API
+directly but uses the BIO abstraction to access the network, so these issues are
+then compounded by the limitations of our existing BIO interfaces. We do not
+have a BIO interface which provides for select(3)-like functionality or which
+can implement the required semantics above. Therefore, trying to implement QUIC
+on top of blocking I/O in this way would require violating the BIO abstraction
+layer, and would not work with custom BIOs.
+
+### Blocking sockets and threads
+
+Another conceptual possibility is that blocking calls could be kept ongoing in
+parallel threads. Under this model, there would be three threads:
+
+- a thread which exists solely to execute blocking calls to the `BIO_write` of
+  an underlying network BIO,
+- a thread which exists solely to execute blocking calls to the `BIO_read` of an
+  underlying network BIO,
+- a thread which exists solely to wait for and dispatch timeout events.
+
+This has a large number of disadvantages:
+
+- There is a hard requirement for threading functionality in order to be
+  able to support blocking semantics at the application level. Use of blocking
+  semantics at the application level will have a hard requirement on use of the
+  thread assisted mode. In environments where threading support is not available
+  or desired, our APIs will only be usable in a non-blocking fashion.
+
+- Several threads are spawned which the application is not in control of.
+  This undermines our general approach of providing the application with control
+  over OpenSSL's use of resources, such as allowing the application to do its
+  own polling or provide its own allocators.
+
+  At a minimum for a client, there must be two threads per connection. This
+  means if an application opens many outgoing connections, there will need
+  to be `2n` extra threads spawned.
+
+- By blocking in `BIO_write` calls, this precludes correct implementation of
+  QUIC. Unlike any analogue in TLS, QUIC packets are time sensitive and intended
+  to be transmitted as soon as they are generated. QUIC packets contain fields
+  such as the ACK Delay value, which is intended to describe the time between a
+  packet being received and a return packet being generated. Correct calculation
+  of this field is necessary to correct calculation of connection RTT. It is
+  therefore important to only generate packets when they are ready to be sent,
+  otherwise suboptimal performance will result. This is a usage model which
+  aligns optimally to non-blocking I/O and which cannot be accommodated
+  by blocking I/O.
+
+- Since existing custom BIOs will not be expecting concurrent `BIO_read` and
+  `BIO_write` calls, they will need to be adapted to support this, which is
+  likely to require substantial rework of those custom BIOs (trivial locking of
+  calls obviously does not work since both of these calls must be able to block
+  on network I/O simultaneously).
```

Comment:
Amongst other reasons ACKs need to be processed, yes.

At any given point in time while we are waiting, *any* of the following things could happen:
- We receive a new packet from the network which contains one or more frames:
  - An ACK frame would be a common frame type to receive, but we could also receive &mdash;
  - `CRYPTO` frames which cause the handshake layer to generate an alert and terminate the connection (unlikely after the handshake is complete, but still possible);
  - `CONNECTION_CLOSE` frames tearing down a connection
  - `STREAM` frames which make a stream readable, and an application on another thread is blocking on a read on that stream (not relevant for MVP but needs to be thought through in advance)
  - etc.
  Moreover incoming packets need to be handled more or less immediately as for most incoming packets we will be expected to generate an ACK promptly in return.
- We become able to write to the network and there is data we would like to be able to send.
  It is important we know we can write to the network before we generate the packet, otherwise
  the data may be stale by the time it is actually transmitted.
- A timeout event occurs. Such events include idle timeout (connection torn down due to lack of traffic, e.g. because the peer has disappeared), ACK generation (which may be delayed for a period), probe generation, or due to the congestion controller now giving permission for more data to be sent, etc.